swe_roe/ex2.c

```diff
@@ -0,0 +1,1224 @@
+  RDyCoordinate *centroid; /* edge centroid                             */
+
+  PetscReal *length; /* length of the edge                        */
+
+} RDyEdge;
+
+typedef struct RDyMesh {
+  PetscInt dim;
+
+  PetscInt num_cells;
+  PetscInt num_cells_local;
+  PetscInt num_edges;
+  PetscInt num_vertices;
+  PetscInt num_boundary_faces;
+
+  PetscInt max_vertex_cells, max_vertex_faces;
+
+  RDyCell   cells;
+  RDyVertex vertices;
+  RDyEdge   edges;
+
+  PetscInt *closureSize, **closure, maxClosureSize;
+
+  PetscInt *nG2A;  // Mapping of global cells to application/natural cells
+
+} RDyMesh;
+
+struct _n_User {
+  MPI_Comm  comm;
+  char      filename[PETSC_MAX_PATH_LEN];
+  DM        dm;
+  PetscInt  Nt, Nx, Ny;
+  PetscReal dt, dx, dy;
+  PetscReal Lx, Ly;
+  PetscReal hu, hd;
+  PetscReal tiny_h;
+  PetscInt  dof, rank, comm_size;
+  Vec       B, localB;
+  Vec       localX;
+  PetscBool debug, save, add_building;
+  PetscInt  tstep;
+  PetscBool interpolate;
+
+  RDyMesh *mesh;
+};
+
+PetscErrorCode RDyInitialize_IntegerArray_1D(PetscInt *array_1D, PetscInt ndim_1, PetscInt init_value) {
+  PetscFunctionBegin;
+  for (PetscInt i = 0; i < ndim_1; i++) {
+    array_1D[i] = init_value;
+  }
+  PetscFunctionReturn(0);
+}
+
+PetscErrorCode RDyAllocate_IntegerArray_1D(PetscInt **array_1D, PetscInt ndim_1) {
+  PetscFunctionBegin;
+  PetscCall(RDyAlloc(ndim_1 * sizeof(PetscInt), array_1D));
+  PetscCall(RDyInitialize_IntegerArray_1D(*array_1D, ndim_1, -1));
+  PetscFunctionReturn(0);
+}
+
+PetscErrorCode RDyAllocate_RDyVector_1D(PetscInt ndim_1, RDyVector **array_1D) { return RDyAlloc(ndim_1 * sizeof(RDyVector), array_1D); }
+
+PetscErrorCode RDyAllocate_RDyCoordinate_1D(PetscInt ndim_1, RDyCoordinate **array_1D) { return RDyAlloc(ndim_1 * sizeof(RDyCoordinate), array_1D); }
+
+PetscErrorCode RDyAllocate_RealArray_1D(PetscReal **array_1D, PetscInt ndim_1) { return RDyAlloc(ndim_1 * sizeof(PetscReal), array_1D); }
+
+PetscBool IsClosureWithinBounds(PetscInt closure, PetscInt start, PetscInt end) { return (closure >= start) && (closure < end); }
+
+/// @brief Process command line options
+/// @param [in] comm A MPI commmunicator
+/// @param [inout] user A User data structure that is updated
+/// @return 0 on success, or a non-zero error code on failure
+static PetscErrorCode ProcessOptions(MPI_Comm comm, User user) {
+  PetscFunctionBegin;
+
+  user->comm   = comm;
+  user->Nx     = 4;
+  user->Ny     = 5;
+  user->dx     = 1.0;
+  user->dy     = 1.0;
+  user->hu     = 10.0;  // water depth for the upstream of dam   [m]
+  user->hd     = 5.0;   // water depth for the downstream of dam [m]
+  user->tiny_h = 1e-7;
+  user->dof    = 3;
+
+  MPI_Comm_size(user->comm, &user->comm_size);
+  MPI_Comm_rank(user->comm, &user->rank);
+
+  PetscErrorCode ierr;
+  ierr = PetscOptionsBegin(user->comm, NULL, "2D Mesh Options", "");
+  PetscCall(ierr);
+  {
+    PetscCall(PetscOptionsInt("-Nx", "Number of cells in X", "", user->Nx, &user->Nx, NULL));
```

Comment:
Are we really supposed to call `PetscCall` for every call to a PETSc function? I wonder if we can configure PETSc to fail on unsuccessful operations instead of having to check `ierr` everywhere. This isn't an issue with this PR, just something I've been curious about lately. I can look more into this.

Reply:
I think we might be able to use [this error handler](https://petsc.org/release/docs/manualpages/Sys/PetscMPIAbortErrorHandler/) to abort when an error is encountered. That would preclude the need for all the error flag checking. Just an idea.

Reply:
You can use the abort handler, but then you don't get a reliable stack trace. Some day we'll use Rust for systems programming and then it's the [`?` operator](https://doc.rust-lang.org/book/ch09-02-recoverable-errors-with-result.html#a-shortcut-for-propagating-errors-the--operator). It's true that stack traces can be obtained using `libunwind`, but PETSc hasn't gone all-in on that route (it adds complexity to the build and may not work everywhere).